### DISCUSSION TITLE
toc shortcode triggering circular references

### DISCUSSION THREAD

**Discussion by jamesbraza** [thegeeklab/hugo-geekdoc#650] [Q&A]
I have the following `content`:

1. `_index.md` refers to:
    - `page1.md` via `{{< ref "page1.md" >}}`
    - `page1.md.part` via `{{< include file="/abs/path/page1.md.part" >}}`
    - `page2.md` via `{{< ref "page2.md" >}}`
    - `page2.md.part` via `{{< include file="/abs/path/page2.md.part" >}}`
2. `page1.md` refers to:
    - `page1.md.part` via `{{< include file="/abs/path/page1.md.part" >}}`
3. `page2.md` refers to:
    - `page2.md.part` via `{{< include file="/abs/path/page2.md.part" >}}`
4. `page1.md.part` refers to nothing
5. `page2.md.part` refers to:
    - `page1.md` via ` {{< ref "page1.md" >}}`

This all works fine, until I add a table of contents via a `toc` shortcode into `_index.md`. The 5th item above triggers a circular reference error.  [This discourse post](https://discourse.gohugo.io/t/you-may-have-a-circular-loop-in-a-shortcode-problem/39183/2) seems relevant.

Any ideas on how to avoid this circular reference?

I also don't quite understand the circular reference, as there's no reference back to `_index.md` in any of the subpages.

---

For a full repro, please check here: https://github.com/jamesbraza/pycon-redux/commit/7ca5edaf128143918c253f9b04ab8775e4458b64

**Comment by xoxys** [comment]
As said in the other discussion, the ToC is a Hugo function and not theme specific. The short code just embeds  `{{ .Page.TableOfContents }}` so this is more a question for the Hugo forum.

**Reply by xoxys** [reply]
It boils down to `content/en/2022/performance-antipatterns.md.part`  and the usage of `{{< ref ... >}}` statements in this file. In general, you create a lot of nested includes and side loads, which makes the entire content structure quite complicated and hard to debug. 

The Table of Content is intended to create a navigation for headers in the *current* page, but you have somewhat created a navigation which jumps to *another* page, IMO this is causing the loop while rendering the ToC.

Don't want to say you can't do it that way, but I can't help much in that case.

**Reply by jamesbraza** [reply]
Thank you for taking a look, I really appreciate it!

It has been interesting, I am going a little wild with the `include` of many `.md.part`s 🤓 . I'm not sure what you mean by a side load.

I am glad newer Hugo versions fixed this, as I think the scenario is a somewhat "downward pointing" dependency tree, not many arrows pointing back "up".

**Comment by jamesbraza** [comment]
Interestingly enough in my playing around, I was testing with Hugo `0.114.0` (current latest release), and it works fine, no circular import issue.

I traced it to between `0.110.0` (circular reference happens) --> `0.111.0` (circular reference doesn't happen).  Reading `0.111.0`'s release notes, it mentions "table of contents" several times, so I guess some new feature resolves this issue.
